R/quosure-join.R

```diff
@@ -0,0 +1,103 @@
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `y` has unique code element placed before common element. This means that `y`
+#' in the environment of the `y` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases to be done
+#' @param x (`Quosure`)
+#' @param y (`Quosure`)
+#' @examples
+#' q1 <- new_quosure(
+#'   code = c(iris1 = "iris1 <- iris", mtcars1 = "mtcars1 <- mtcars"),
+#'   env = list2env(list(
+#'     iris1 = iris,
+#'     mtcars1 = mtcars
+#'   ))
+#' )
+#' q2 <- q1
+#' q1 <- eval_code(q1, "iris2 <- iris")
+#' q2 <- eval_code(q2, "mtcars2 <- mtcars")
+#' qq <- join(q1, q2)
+#' get_code(qq)
+#' @export
+setGeneric("join", function(x, y) {
+  standardGeneric("join")
+})
+
+#' @rdname join
+#' @export
+setMethod("join", signature = c("Quosure", "Quosure"), function(x, y) {
+  join_validation <- check_joinable(x, y)
+
+  # join expressions
+  if (!isTRUE(join_validation)) {
+    stop(join_validation)
+  }
+
+  id_unique <- !y@id %in% x@id
+  x@id <- c(x@id, y@id[id_unique])
+  x@code <- .keep_code_name_unique(x@code, y@code[id_unique])
+
+  # insert (and overwrite) objects from env2 to env
+  x@env <- .copy_env(x@env)
+  lapply(ls(y@env, all.names = TRUE), function(xx) assign(xx, get(xx, y@env), x@env))
```

Comment:
[`rlang::env_coalesce`](https://rlang.r-lib.org/reference/env_clone.html)

Reply:
I think you are right here and comment from 44 is not valid as `check_joinable` prevent to join when one quosure modified any of the common objects.

Reply:
marking as resolved 
- PR here: https://github.com/insightsengineering/teal.code/pull/68